// @graphql-modules/epoxy/dist/schema-mergers/type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fields_1 = require("./fields");
var directives_1 = require("./directives");
var merge_named_type_array_1 = require("./merge-named-type-array");
function mergeType(node, existingNode) {
    if (existingNode) {
        try {
            return {
                name: node.name,
                description: node['description'] || existingNode['description'],
                kind: (node.kind === 'ObjectTypeDefinition' || existingNode.kind === 'ObjectTypeDefinition') ? 'ObjectTypeDefinition' : 'ObjectTypeExtension',
                loc: node.loc,
                fields: fields_1.mergeFields(node.fields, existingNode.fields),
                directives: directives_1.mergeDirectives(node.directives, existingNode.directives),
                interfaces: merge_named_type_array_1.mergeNamedTypeArray(node.interfaces, existingNode.interfaces),
            };
        }
        catch (e) {
            throw new Error("Unable to merge GraphQL type \"" + node.name.value + "\": " + e.message);
        }
    }
    return node;
}
exports.mergeType = mergeType;
