// @graphql-modules/epoxy/dist/schema-mergers/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_1 = require("graphql");
function isStringTypes(types) {
    return typeof types === 'string';
}
exports.isStringTypes = isStringTypes;
function isSourceTypes(types) {
    return types instanceof graphql_1.Source;
}
exports.isSourceTypes = isSourceTypes;
function isGraphQLType(definition) {
    return definition.kind === 'ObjectTypeDefinition';
}
exports.isGraphQLType = isGraphQLType;
function isGraphQLTypeExtension(definition) {
    return definition.kind === 'ObjectTypeExtension';
}
exports.isGraphQLTypeExtension = isGraphQLTypeExtension;
function isGraphQLEnum(definition) {
    return definition.kind === 'EnumTypeDefinition';
}
exports.isGraphQLEnum = isGraphQLEnum;
function isGraphQLEnumExtension(definition) {
    return definition.kind === 'EnumTypeExtension';
}
exports.isGraphQLEnumExtension = isGraphQLEnumExtension;
function isGraphQLUnion(definition) {
    return definition.kind === 'UnionTypeDefinition';
}
exports.isGraphQLUnion = isGraphQLUnion;
function isGraphQLUnionExtension(definition) {
    return definition.kind === 'UnionTypeExtension';
}
exports.isGraphQLUnionExtension = isGraphQLUnionExtension;
function isGraphQLScalar(definition) {
    return definition.kind === 'ScalarTypeDefinition';
}
exports.isGraphQLScalar = isGraphQLScalar;
function isGraphQLScalarExtension(definition) {
    return definition.kind === 'ScalarTypeExtension';
}
exports.isGraphQLScalarExtension = isGraphQLScalarExtension;
function isGraphQLInputType(definition) {
    return definition.kind === 'InputObjectTypeDefinition';
}
exports.isGraphQLInputType = isGraphQLInputType;
function isGraphQLInputTypeExtension(definition) {
    return definition.kind === 'InputObjectTypeExtension';
}
exports.isGraphQLInputTypeExtension = isGraphQLInputTypeExtension;
function isGraphQLInterface(definition) {
    return definition.kind === 'InterfaceTypeDefinition';
}
exports.isGraphQLInterface = isGraphQLInterface;
function isGraphQLInterfaceExtension(definition) {
    return definition.kind === 'InterfaceTypeExtension';
}
exports.isGraphQLInterfaceExtension = isGraphQLInterfaceExtension;
function isGraphQLDirective(definition) {
    return definition.kind === 'DirectiveDefinition';
}
exports.isGraphQLDirective = isGraphQLDirective;
function isGraphQLSchema(obj) {
    return obj instanceof graphql_1.GraphQLSchema;
}
exports.isGraphQLSchema = isGraphQLSchema;
function extractType(type) {
    var visitedType = type;
    while (visitedType.kind === 'ListType' || visitedType.kind === 'NonNullType') {
        visitedType = visitedType.type;
    }
    return visitedType;
}
exports.extractType = extractType;
function isSchemaDefinition(node) {
    return node.kind === 'SchemaDefinition';
}
exports.isSchemaDefinition = isSchemaDefinition;
