// apollo-codegen/lib/swift/codeGeneration.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _entries = require('babel-runtime/core-js/object/entries');

var _entries2 = _interopRequireDefault(_entries);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _values = require('babel-runtime/core-js/object/values');

var _values2 = _interopRequireDefault(_values);

exports.generateSource = generateSource;
exports.classDeclarationForOperation = classDeclarationForOperation;
exports.initializerDeclarationForProperties = initializerDeclarationForProperties;
exports.structDeclarationForFragment = structDeclarationForFragment;
exports.structDeclarationForSelectionSet = structDeclarationForSelectionSet;
exports.initializationForProperty = initializationForProperty;
exports.dictionaryLiteralForFieldArguments = dictionaryLiteralForFieldArguments;
exports.propertiesFromFields = propertiesFromFields;
exports.propertyFromField = propertyFromField;
exports.structNameForProperty = structNameForProperty;
exports.typeNameForFragmentName = typeNameForFragmentName;
exports.typeDeclarationForGraphQLType = typeDeclarationForGraphQLType;

var _graphql = require('graphql');

var _graphql2 = require('../utilities/graphql');

var _changeCase = require('change-case');

var _inflected = require('inflected');

var _inflected2 = _interopRequireDefault(_inflected);

var _printing = require('../utilities/printing');

var _language = require('./language');

var _values3 = require('./values');

var _types = require('./types');

var _CodeGenerator = require('../utilities/CodeGenerator');

var _CodeGenerator2 = _interopRequireDefault(_CodeGenerator);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function generateSource(context) {
  var generator = new _CodeGenerator2.default(context);

  generator.printOnNewline('//  This file was automatically generated and should not be edited.');
  generator.printNewline();
  generator.printOnNewline('import Apollo');

  context.typesUsed.forEach(function (type) {
    typeDeclarationForGraphQLType(generator, type);
  });

  (0, _values2.default)(context.operations).forEach(function (operation) {
    classDeclarationForOperation(generator, operation);
  });

  (0, _values2.default)(context.fragments).forEach(function (fragment) {
    structDeclarationForFragment(generator, fragment);
  });

  return generator.output;
}

function classDeclarationForOperation(generator, _ref) {
  var operationName = _ref.operationName,
      operationType = _ref.operationType,
      variables = _ref.variables,
      fields = _ref.fields,
      fragmentsReferenced = _ref.fragmentsReferenced,
      source = _ref.source;


  var className = void 0;
  var protocol = void 0;

  switch (operationType) {
    case 'query':
      className = `${(0, _changeCase.pascalCase)(operationName)}Query`;
      protocol = 'GraphQLQuery';
      break;
    case 'mutation':
      className = `${(0, _changeCase.pascalCase)(operationName)}Mutation`;
      protocol = 'GraphQLMutation';
      break;
    default:
      throw new _graphql.GraphQLError(`Unsupported operation type "${operationType}"`);
  }

  (0, _language.classDeclaration)(generator, {
    className,
    modifiers: ['public', 'final'],
    adoptedProtocols: [protocol]
  }, function () {
    if (source) {
      generator.printOnNewline('public static let operationDefinition =');
      generator.withIndent(function () {
        (0, _values3.multilineString)(generator, source);
      });
    }

    if (fragmentsReferenced && fragmentsReferenced.length > 0) {
      generator.printOnNewline('public static let queryDocument = operationDefinition');
      fragmentsReferenced.forEach(function (fragment) {
        generator.print(`.appending(${typeNameForFragmentName(fragment)}.fragmentDefinition)`);
      });
    }

    if (variables && variables.length > 0) {
      var properties = variables.map(function (_ref2) {
        var name = _ref2.name,
            type = _ref2.type;

        var propertyName = (0, _language.escapeIdentifierIfNeeded)((0, _changeCase.camelCase)(name));
        var typeName = (0, _types.typeNameFromGraphQLType)(generator.context, type);
        var isOptional = !(type instanceof _graphql.GraphQLNonNull || type.ofType instanceof _graphql.GraphQLNonNull);
        return { name, propertyName, type, typeName, isOptional };
      });
      generator.printNewlineIfNeeded();
      (0, _language.propertyDeclarations)(generator, properties);
      generator.printNewlineIfNeeded();
      initializerDeclarationForProperties(generator, properties);
      generator.printNewlineIfNeeded();
      generator.printOnNewline(`public var variables: GraphQLMap?`);
      generator.withinBlock(function () {
        generator.printOnNewline((0, _printing.wrap)(`return [`, (0, _printing.join)(properties.map(function (_ref3) {
          var name = _ref3.name,
              propertyName = _ref3.propertyName;
          return `"${name}": ${propertyName}`;
        }), ', ') || ':', `]`));
      });
    } else {
      initializerDeclarationForProperties(generator, []);
    }

    structDeclarationForSelectionSet(generator, {
      structName: "Data",
      fields
    });
  });
}

function initializerDeclarationForProperties(generator, properties) {
  generator.printOnNewline(`public init`);
  generator.print('(');
  generator.print((0, _printing.join)(properties.map(function (_ref4) {
    var propertyName = _ref4.propertyName,
        type = _ref4.type,
        typeName = _ref4.typeName,
        isOptional = _ref4.isOptional;
    return (0, _printing.join)([`${propertyName}: ${typeName}`, isOptional && ' = nil']);
  }), ', '));
  generator.print(')');

  generator.withinBlock(function () {
    properties.forEach(function (_ref5) {
      var propertyName = _ref5.propertyName;

      generator.printOnNewline(`self.${propertyName} = ${propertyName}`);
    });
  });
}

function structDeclarationForFragment(generator, _ref6) {
  var fragmentName = _ref6.fragmentName,
      typeCondition = _ref6.typeCondition,
      possibleTypes = _ref6.possibleTypes,
      fields = _ref6.fields,
      inlineFragments = _ref6.inlineFragments,
      fragmentSpreads = _ref6.fragmentSpreads,
      source = _ref6.source;

  var structName = (0, _changeCase.pascalCase)(fragmentName);

  structDeclarationForSelectionSet(generator, {
    structName,
    adoptedProtocols: ['GraphQLNamedFragment'],
    parentType: typeCondition,
    possibleTypes,
    fields,
    fragmentSpreads,
    inlineFragments
  }, function () {
    if (source) {
      generator.printOnNewline('public static let fragmentDefinition =');
      generator.withIndent(function () {
        (0, _values3.multilineString)(generator, source);
      });
    }
  });
}

function structDeclarationForSelectionSet(generator, _ref7, beforeClosure) {
  var structName = _ref7.structName,
      _ref7$adoptedProtocol = _ref7.adoptedProtocols,
      adoptedProtocols = _ref7$adoptedProtocol === undefined ? ['GraphQLMappable'] : _ref7$adoptedProtocol,
      parentType = _ref7.parentType,
      possibleTypes = _ref7.possibleTypes,
      fields = _ref7.fields,
      fragmentSpreads = _ref7.fragmentSpreads,
      inlineFragments = _ref7.inlineFragments;

  (0, _language.structDeclaration)(generator, { structName, adoptedProtocols }, function () {
    if (beforeClosure) {
      beforeClosure();
    }

    if (possibleTypes) {
      generator.printNewlineIfNeeded();
      generator.printOnNewline('public static let possibleTypes = [');
      generator.print((0, _printing.join)(possibleTypes.map(function (type) {
        return `"${String(type)}"`;
      }), ', '));
      generator.print(']');
    }

    var properties = fields && propertiesFromFields(generator.context, fields);

    var fragmentProperties = fragmentSpreads && fragmentSpreads.map(function (fragmentName) {
      var fragment = generator.context.fragments[fragmentName];
      if (!fragment) {
        throw new _graphql.GraphQLError(`Cannot find fragment "${fragmentName}"`);
      }
      var propertyName = (0, _changeCase.camelCase)(fragmentName);
      var typeName = typeNameForFragmentName(fragmentName);
      var isProperSuperType = (0, _graphql2.isTypeProperSuperTypeOf)(generator.context.schema, fragment.typeCondition, parentType);
      return { propertyName, typeName, bareTypeName: typeName, isProperSuperType };
    });

    var inlineFragmentProperties = inlineFragments && inlineFragments.map(function (inlineFragment) {
      var bareTypeName = 'As' + (0, _changeCase.pascalCase)(String(inlineFragment.typeCondition));
      var propertyName = (0, _changeCase.camelCase)(bareTypeName);
      var typeName = bareTypeName + '?';
      return (0, _extends3.default)({}, inlineFragment, { propertyName, typeName, bareTypeName });
    });

    generator.printNewlineIfNeeded();

    if (parentType) {
      generator.printOnNewline('public let __typename: String');
    }

    (0, _language.propertyDeclarations)(generator, properties);

    if (fragmentProperties && fragmentProperties.length > 0) {
      generator.printNewlineIfNeeded();
      (0, _language.propertyDeclaration)(generator, { propertyName: 'fragments', typeName: 'Fragments' });
    }

    if (inlineFragmentProperties && inlineFragmentProperties.length > 0) {
      generator.printNewlineIfNeeded();
      (0, _language.propertyDeclarations)(generator, inlineFragmentProperties);
    }

    generator.printNewlineIfNeeded();
    generator.printOnNewline('public init(reader: GraphQLResultReader) throws');
    generator.withinBlock(function () {
      if (parentType) {
        generator.printOnNewline(`__typename = try reader.value(for: Field(responseName: "__typename"))`);
      }

      if (properties) {
        properties.forEach(function (property) {
          return initializationForProperty(generator, property);
        });
      }

      if (fragmentProperties && fragmentProperties.length > 0) {
        generator.printNewlineIfNeeded();
        fragmentProperties.forEach(function (_ref8) {
          var propertyName = _ref8.propertyName,
              typeName = _ref8.typeName,
              bareTypeName = _ref8.bareTypeName,
              isProperSuperType = _ref8.isProperSuperType;

          generator.printOnNewline(`let ${propertyName} = try ${typeName}(reader: reader`);
          if (isProperSuperType) {
            generator.print(')');
          } else {
            generator.print(`, ifTypeMatches: __typename)`);
          }
        });
        generator.printOnNewline(`fragments = Fragments(`);
        generator.print((0, _printing.join)(fragmentSpreads.map(function (fragmentName) {
          var propertyName = (0, _changeCase.camelCase)(fragmentName);
          return `${propertyName}: ${propertyName}`;
        }), ', '));
        generator.print(')');
      }

      if (inlineFragmentProperties && inlineFragmentProperties.length > 0) {
        generator.printNewlineIfNeeded();
        inlineFragmentProperties.forEach(function (_ref9) {
          var propertyName = _ref9.propertyName,
              typeName = _ref9.typeName,
              bareTypeName = _ref9.bareTypeName;

          generator.printOnNewline(`${propertyName} = try ${bareTypeName}(reader: reader, ifTypeMatches: __typename)`);
        });
      }
    });

    if (fragmentProperties && fragmentProperties.length > 0) {
      (0, _language.structDeclaration)(generator, {
        structName: 'Fragments'
      }, function () {
        fragmentProperties.forEach(function (_ref10) {
          var propertyName = _ref10.propertyName,
              typeName = _ref10.typeName,
              isProperSuperType = _ref10.isProperSuperType;

          if (!isProperSuperType) {
            typeName += '?';
          }
          (0, _language.propertyDeclaration)(generator, { propertyName, typeName });
        });
      });
    }

    if (inlineFragmentProperties && inlineFragmentProperties.length > 0) {
      inlineFragmentProperties.forEach(function (property) {
        structDeclarationForSelectionSet(generator, {
          structName: property.bareTypeName,
          parentType: property.typeCondition,
          possibleTypes: property.possibleTypes,
          adoptedProtocols: ['GraphQLConditionalFragment'],
          fields: property.fields,
          fragmentSpreads: property.fragmentSpreads
        });
      });
    }

    if (properties) {
      properties.filter(function (property) {
        return property.isComposite;
      }).forEach(function (property) {
        structDeclarationForSelectionSet(generator, {
          structName: structNameForProperty(property),
          parentType: (0, _graphql.getNamedType)(property.type),
          fields: property.fields,
          fragmentSpreads: property.fragmentSpreads,
          inlineFragments: property.inlineFragments
        });
      });
    }
  });
}

function initializationForProperty(generator, _ref11) {
  var propertyName = _ref11.propertyName,
      responseName = _ref11.responseName,
      fieldName = _ref11.fieldName,
      fieldArgs = _ref11.args,
      type = _ref11.type,
      isOptional = _ref11.isOptional;

  var isList = type instanceof _graphql.GraphQLList || type.ofType instanceof _graphql.GraphQLList;

  var methodName = isOptional ? isList ? 'optionalList' : 'optionalValue' : isList ? 'list' : 'value';

  var fieldInitArgs = (0, _printing.join)([`responseName: "${responseName}"`, responseName != fieldName ? `fieldName: "${fieldName}"` : null, fieldArgs && fieldArgs.length && `arguments: ${dictionaryLiteralForFieldArguments(fieldArgs)}`], ', ');
  var args = [`for: Field(${fieldInitArgs})`];

  generator.printOnNewline(`${propertyName} = try reader.${methodName}(${(0, _printing.join)(args, ', ')})`);
}

function dictionaryLiteralForFieldArguments(args) {
  function expressionFromValue(value) {
    if (value.kind === 'Variable') {
      return `reader.variables["${value.variableName}"]`;
    } else if (Array.isArray(value)) {
      return (0, _printing.wrap)('[', (0, _printing.join)(value.map(expressionFromValue), ', '), ']');
    } else if (typeof value === 'object') {
      return (0, _printing.wrap)('[', (0, _printing.join)((0, _entries2.default)(value).map(function (_ref12) {
        var _ref13 = (0, _slicedToArray3.default)(_ref12, 2),
            key = _ref13[0],
            value = _ref13[1];

        return `"${key}": ${expressionFromValue(value)}`;
      }), ', ') || ':', ']');
    } else {
      return (0, _stringify2.default)(value);
    }
  }

  return (0, _printing.wrap)('[', (0, _printing.join)(args.map(function (arg) {
    return `"${arg.name}": ${expressionFromValue(arg.value)}`;
  }), ', ') || ':', ']');
}

function propertiesFromFields(context, fields) {
  return fields.map(function (field) {
    return propertyFromField(context, field);
  });
}

function propertyFromField(context, field) {
  var name = field.name || field.responseName;
  var propertyName = (0, _language.escapeIdentifierIfNeeded)((0, _changeCase.camelCase)(name));

  var type = field.type;
  var isOptional = field.isConditional || !(type instanceof _graphql.GraphQLNonNull);
  var bareType = (0, _graphql.getNamedType)(type);

  if ((0, _graphql.isCompositeType)(bareType)) {
    var bareTypeName = (0, _language.escapeIdentifierIfNeeded)((0, _changeCase.pascalCase)(_inflected2.default.singularize(name)));
    var typeName = (0, _types.typeNameFromGraphQLType)(context, type, bareTypeName, isOptional);
    return (0, _extends3.default)({}, field, { propertyName, typeName, bareTypeName, isOptional, isComposite: true });
  } else {
    var _typeName = (0, _types.typeNameFromGraphQLType)(context, type, undefined, isOptional);
    return (0, _extends3.default)({}, field, { propertyName, typeName: _typeName, isOptional, isComposite: false });
  }
}

function structNameForProperty(property) {
  return (0, _changeCase.pascalCase)(_inflected2.default.singularize(property.responseName));
}

function typeNameForFragmentName(fragmentName) {
  return (0, _changeCase.pascalCase)(fragmentName);
}

function typeDeclarationForGraphQLType(generator, type) {
  if (type instanceof _graphql.GraphQLEnumType) {
    enumerationDeclaration(generator, type);
  } else if (type instanceof _graphql.GraphQLInputObjectType) {
    structDeclarationForInputObjectType(generator, type);
  }
}

function enumerationDeclaration(generator, type) {
  var name = type.name,
      description = type.description;

  var values = type.getValues();

  generator.printNewlineIfNeeded();
  generator.printOnNewline(description && `/// ${description}`);
  generator.printOnNewline(`public enum ${name}: String`);
  generator.withinBlock(function () {
    values.forEach(function (value) {
      return generator.printOnNewline(`case ${(0, _language.escapeIdentifierIfNeeded)((0, _changeCase.camelCase)(value.name))} = "${value.value}"${(0, _printing.wrap)(' /// ', value.description)}`);
    });
  });
  generator.printNewline();
  generator.printOnNewline(`extension ${name}: JSONDecodable, JSONEncodable {}`);
}

function structDeclarationForInputObjectType(generator, type) {
  var structName = type.name,
      description = type.description;

  var adoptedProtocols = ['GraphQLMapConvertible'];
  var properties = propertiesFromFields(generator.context, (0, _values2.default)(type.getFields()));

  (0, _language.structDeclaration)(generator, { structName, description, adoptedProtocols }, function () {
    generator.printOnNewline(`public var graphQLMap: GraphQLMap`);

    generator.printNewlineIfNeeded();
    generator.printOnNewline(`public init`);
    generator.print('(');
    generator.print((0, _printing.join)(properties.map(function (_ref14) {
      var propertyName = _ref14.propertyName,
          type = _ref14.type,
          typeName = _ref14.typeName,
          isOptional = _ref14.isOptional;
      return (0, _printing.join)([`${propertyName}: ${typeName}`, isOptional && ' = nil']);
    }), ', '));
    generator.print(')');

    generator.withinBlock(function () {
      generator.printOnNewline((0, _printing.wrap)(`graphQLMap = [`, (0, _printing.join)(properties.map(function (_ref15) {
        var name = _ref15.name,
            propertyName = _ref15.propertyName;
        return `"${name}": ${propertyName}`;
      }), ', ') || ':', `]`));
    });
  });
}
