// apollo-language-server/lib/fileSet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const minimatch = require("minimatch");
const glob = require("glob");
const apollo_tools_1 = require("@apollographql/apollo-tools");
class FileSet {
    constructor({ rootURI, includes, excludes }) {
        apollo_tools_1.invariant(rootURI, `Must provide "rootURI".`);
        apollo_tools_1.invariant(includes, `Must provide "includes".`);
        apollo_tools_1.invariant(excludes, `Must provide "excludes".`);
        this.rootURI = rootURI;
        this.includes = includes;
        this.excludes = excludes;
    }
    includesFile(filePath) {
        filePath = path_1.relative(this.rootURI.fsPath, filePath);
        return (this.includes.some(include => minimatch(filePath, include)) &&
            !this.excludes.some(exclude => minimatch(filePath, exclude)));
    }
    allFiles() {
        return this.includes
            .flatMap(include => glob.sync(include, { cwd: this.rootURI.fsPath, absolute: true }))
            .filter(filePath => !this.excludes.some(exclude => minimatch(path_1.relative(this.rootURI.fsPath, filePath), exclude)));
    }
}
exports.FileSet = FileSet;
