// graphql-codegen-plugin-helpers/dist/build-files-array.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var path_1 = require("path");
function getFileAttrs(filePath) {
    var cwd = process.cwd();
    return {
        cwd: cwd,
        relative: filePath,
        absolute: path_1.resolve(filePath),
        filename: path_1.basename(filePath),
        extension: path_1.extname(filePath)
    };
}
function buildFilesArray(parsedDocument) {
    var filesMap = {};
    var relevantOperations = parsedDocument.operations.concat(parsedDocument.fragments).filter(function (o) { return o.originalFile; });
    for (var _i = 0, relevantOperations_1 = relevantOperations; _i < relevantOperations_1.length; _i++) {
        var operation = relevantOperations_1[_i];
        if (!filesMap[operation.originalFile]) {
            filesMap[operation.originalFile] = [];
        }
        filesMap[operation.originalFile].push(operation);
    }
    return Object.keys(filesMap).map(function (filename) { return (__assign({ documents: filesMap[filename] }, getFileAttrs(filename))); });
}
exports.buildFilesArray = buildFilesArray;
